canonical/main.go

```diff
@@ -0,0 +1,377 @@
+
+	if *memprofile != "" {
+		f, err := os.Create(*memprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		runtime.GC() // get up-to-date statistics
+		pprof.WriteHeapProfile(f)
+		f.Close()
+		return
+	}
+
+	if err != nil {
+		fmt.Println("error processing bucket:", err)
+		return
+	}
+}
+
+// processBucket processes all tarballs in the given bucket
+// and stores the output in the output bucket.
+// If processEverything is false, only new tarballs will be processed.
+func processBucket(ctx context.Context, inBucketName, outBucketName string, processEverything bool) error {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+
+	inBucket := client.Bucket(inBucketName)
+	if _, err := inBucket.Attrs(ctx); err != nil {
+		return fmt.Errorf("could not get bucket %q: %v", inBucketName, err)
+	}
+
+	outBucket := client.Bucket(outBucketName)
+	if _, err := outBucket.Attrs(ctx); err != nil {
+		return fmt.Errorf("could not get bucket %q: %v", outBucketName, err)
+	}
+
+	allInputs, err := objects(ctx, inBucket)
+	if err != nil {
+		return err
+	}
+
+	existingOutputs, err := objects(ctx, outBucket)
+	if err != nil {
+		return err
+	}
+
+	inputs := []string{}
+	for name := range allInputs {
+		if processEverything || !existingOutputs[name] {
+			inputs = append(inputs, name)
+		}
+	}
+
+	if len(inputs) == 0 {
+		fmt.Println("Nothing to do. Add -f flag?")
+		return nil
+	}
+
+	bar := pbTemplate.Start(len(inputs)).Set("prefix", "Canonicalizing... ").SetRefreshRate(time.Second)
+
+	g, ctx := errgroup.WithContext(ctx)
+	sem := semaphore.NewWeighted(500)
+
+	for _, name := range inputs {
+		if err := ctx.Err(); ctx.Err() != nil {
+			bar.Finish()
+			log.Println(err)
+			break
+		}
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			bar.Finish()
+			log.Println(err)
+			break
+		}
+		releaseOnce := &sync.Once{}
+
+		// Capture for loop variable.
+		name := name
+
+		g.Go(func() error {
+			defer bar.Increment()
+			defer releaseOnce.Do(func() { sem.Release(1) })
+
+			return processTarball(ctx, inBucket, outBucket, name, bar)
+		})
+	}
+	if err := g.Wait(); err != nil {
+		bar.Finish()
+		return err
+	}
+	bar.Finish()
+
+	return nil
+}
+
+var numFiles int64
+
+// processTarball processes the given tarball in the given bucket.
+// The progress bar is updated with the number of files processed as a prefix.
+func processTarball(ctx context.Context, inBucket, outBucket *storage.BucketHandle, name string, bar *pb.ProgressBar) error {
+	pkg := parsePkg(name)
+
+	obj := inBucket.Object(name)
+
+	// Storage reader -> un-gzip -> un-tar
+	// -> process
+	// -> tar -> gzip -> Storage writer.
+	r, err := obj.NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to read from gs://%s/%s: %v", obj.BucketName(), name, err)
+	}
+
+	gr, err := gzip.NewReader(r)
+	if err != nil {
+		log.Fatal(err)
+	}
+	tr := tar.NewReader(gr)
+
+	outObj := outBucket.Object(name)
+	ow := outObj.NewWriter(ctx)
```

Comment:
In addition to the explicit Closes later should these be defered as well for the error cases where the loop returns?

Reply:
Done. Just removed the explicit closes.